tests/websocket.c

```diff
@@ -8,319 +8,537 @@
+
+    /* getopt */
+    struct option long_options[] = {
+        {"help", no_argument, NULL, '?'},
+        {"host", required_argument, NULL, 'h'},
+        {"port", required_argument, NULL, 'p'},
+        {"clients", required_argument, NULL, 'c'},
+        {"messages", required_argument, NULL, 'n'},
+        {"max-time", required_argument, NULL, 't'},
+        {"verbose", no_argument, NULL, 'v'},
+        {0, 0, 0, 0}
+    };
+    while ((opt = getopt_long(argc, argv, "h:p:c:n:t:vs", long_options, NULL)) != -1) {
+        switch (opt) {
+            case 'h':
+                colon = strchr(optarg, ':');
+                if(!colon) {
+                    size_t sz = strlen(optarg);
+                    hi.host = calloc(1 + sz, 1);
+                    strncpy(hi.host, optarg, sz);
+                } else {
+                    hi.host = calloc(1+colon-optarg, 1);
+                    strncpy(hi.host, optarg, colon-optarg);
+                    hi.port = (short)atol(colon+1);
+                }
+                break;
+
+            case 'p':
+                hi.port = (short)atol(optarg);
+                break;
+
+            case 'c':
+                thread_count = atoi(optarg);
+                break;
+
+            case 'n':
+                msg_target = atoi(optarg);
+                break;
+
+            case 't':
+                timeout_seconds = atoi(optarg);
+                break;
+
+            case 'v':
+                verbose = 1;
+                break;
+
+            default: /* '?' */
+                usage(argv[0], host_default, port_default,
+                        thread_count_default,
+                        messages_default);
+                exit(EXIT_SUCCESS);
+        }
+    }
+
+    /* run threads */
+    workers = calloc(sizeof(struct worker_thread), thread_count);
+
+    clock_gettime(CLOCK_MONOTONIC, &t0);
+    for (i = 0; i < thread_count; ++i) {
+        workers[i].id = i;
+        workers[i].msg_target = msg_target;
+        workers[i].hi = &hi;
+        workers[i].verbose = verbose;
+        workers[i].state = WS_INITIAL;
+        workers[i].debug = verbose ? debug_verbose : debug_noop;
+        workers[i].timeout_seconds = timeout_seconds;
+        if (thread_count == 1) {
+            printf("Single-threaded mode\n");
+            worker_main(&workers[0]);
+        } else { /* create threads */
+            pthread_create(&workers[i].thread, NULL,
+                           worker_main, &workers[i]);
+        }
+    }
+
+    /* wait for threads to finish */
+    for (i = 0; i < thread_count; ++i) {
+        if (thread_count > 1) {
+            pthread_join(workers[i].thread, NULL);
+        }
+        total += workers[i].msg_received;
+        total_bytes += workers[i].byte_count;
+    }
+
+    /* timing */
+    clock_gettime(CLOCK_MONOTONIC, &t1);
+    float mili0 = t0.tv_sec * 1000 + t0.tv_nsec / 1000000;
+    float mili1 = t1.tv_sec * 1000 + t1.tv_nsec / 1000000;
+
+    if(total != 0) {
```

Comment:
minor, but related to the spacing comment at the top of this file: most `if` blocks in the code base don't have space after `if`. If this much code is changed then I think it's fine to have a space, but let's at least keep consistent.